libcloud/common/exception.py

```diff
@@ -0,0 +1,66 @@
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+class BaseHTTPException(Exception):
+
+    """
+    The base exception class for all exception raises.
+    """
+
+    def __init__(self, code, message, headers=None):
+
+        self.message = message
+        self.code = code
+        self.headers = headers
+
+
+class RateLimit(BaseHTTPException):
+    """
+    HTTP 429 - Rate limit: you've sent too many requests for this time period.
+    """
+    http_status = 429
+    message = "Rate limit"
+
+    def __init__(self, *args, **kwargs):
+        try:
+            self.retry_after = int(kwargs.pop('retry_after'))
+        except (KeyError, ValueError):
+            self.retry_after = 0
+
+
+_error_classes = [RateLimit]
+_code_map = dict((c.http_status, c) for c in _error_classes)
+
+
+def exception_from_message(code, message, headers=None):
+    """
+    Return an instance of BaseHTTPException or subclass based on response code.
+
+    Usage::
+        raise exception_from_message(code=self.status,
+                                     message=self.parse_error())
+    """
+    kwargs = {
+        'code': code,
+        'message': message,
+        'headers': headers
+    }
+
+    if headers:
+        kwargs['retry_after'] = headers['retry_after']
+
+    cls = _code_map.get(code, BaseHTTPException)
+    return cls(**kwargs)
```

Comment:
Missing final newline.  Make sure you run ```tox -e lint``` against your code if you haven't already.